lxl-web/src/lib/utils/auxd.ts

```diff
@@ -1,51 +1,71 @@
+	type ImageResolution,
+	isImage,
+	isImageResolution
+} from '$lib/utils/auxd.types';
+import getAtPath from '$lib/utils/getAtPath';
+import { relativizeUrl, stripAnchor } from '$lib/utils/http';
 
-export function getImageLinks(mainEntity) {
-	const instances = getInstances(mainEntity);
-	if (Object.hasOwn(mainEntity, 'image')) {
-		return [
-			{
-				recordId: mainEntity['@id'],
-				imageLink: asArray(mainEntity.image)[0]['@id']
-			}
-		];
-	} else if (instances?.length) {
-		const imageLinks = instances.map((i) => {
-			if (i.image) {
-				return {
-					recordId: i['@id'],
-					imageLink: asArray(i.image)[0]['@id']
-				};
-			} else {
-				return {
-					recordId: i['@id'],
-					imageLink: ''
-				};
-			}
-		});
-		return imageLinks;
-	} else {
-		return [];
+function toImage(imageObject: KbvImageObject, recordId: string): Image {
+	const map1 = (i: KbvImageObject) =>
+		({
+			url: i[JsonLd.ID] || i[Owl.SAME_AS][0][JsonLd.ID],
+			widthṔx: Number.parseInt(i.width || ''),
+			heightPx: Number.parseInt(i.height || '')
+		}) as ImageResolution;
+
+	const sizes = [...(imageObject.thumbnail?.map(map1) || []), map1(imageObject)];
+	sizes.sort((a, b) => a.widthṔx - b.widthṔx);
+	return { sizes: sizes, recordId: recordId };
+}
+
+export function bestSize(from: Image, minWidthPx: number): ImageResolution;
+export function bestSize(from: undefined, minWidthPx: number): undefined;
+export function bestSize(from: Image | undefined, minWidthPx: number): ImageResolution | undefined {
+	if (from === undefined) {
+		return undefined;
 	}
+	const sizes = from.sizes;
+	const result = sizes.find((i) => i.widthṔx >= minWidthPx);
+	return result || sizes[sizes.length - 1];
+}
+
+export function getImages(thing: FramedData): Image[] {
+	return [
+		...asArray(thing.image).map((i) =>
+			toImage(i as KbvImageObject, stripAnchor(relativizeUrl(thing['@id']) as string))
+		),
+		...getInstances(thing).flatMap(getImages)
+	];
 }
 
-export function getFirstImageLink(mainEntity) {
-	const links = getImageLinks(mainEntity);
-	const nonEmptyLinks = links.filter((l) => l.imageLink !== '');
-	return nonEmptyLinks.length ? nonEmptyLinks[0].imageLink : '';
+export function bestImage(thing: FramedData): Image | undefined {
+	return first(getImages(thing));
 }
 
-export function getFirstImageUri(uris) {
-	const nonEmptyUris = uris.filter((u) => u.imageUri !== '');
-	return nonEmptyUris.length ? nonEmptyUris[0].imageUri : '';
+function getInstances(thing: FramedData): FramedData[] {
+	return getAtPath(thing, ['@reverse', 'instanceOf', '*'], []);
 }
 
-function getInstances(mainEntity) {
-	if (mainEntity['@reverse'] && mainEntity['@reverse']['instanceOf']) {
-		return mainEntity['@reverse']['instanceOf'];
-	} else {
-		return [];
+export function auxdAuth(i: undefined, secret: string): undefined;
+export function auxdAuth(i: Image, secret: string): AuthImage;
+export function auxdAuth(i: ImageResolution, secret: string): AuthImageResolution;
+export function auxdAuth(
```

Comment:
Maybe come upp with a more descriptive name for this function?

Reply:
Trying with the "secure link" terminology from nginx